prov/verbs_ud/src/verbs_ud.h

```diff
@@ -0,0 +1,738 @@
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifndef _OFI_VERBS_UD_H_
+#define _OFI_VERBS_UD_H_
+
+#if HAVE_CONFIG_H
+#  include <config.h>
+#endif /* HAVE_CONFIG_H */
+
+#include <infiniband/verbs.h>
+#include <rdma/rdma_cma.h>
+
+#include <fi_util.h>
+
+#include <inttypes.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif /* __cplusplus */
+
+#define OFI_VERBS_UD_MAJOR_VERSION 1
+#define OFI_VERBS_UD_MINOR_VERSION 5
+
+#define OFI_VERBS_UD_PRIMARY_CAPS (FI_MSG | FI_RMA |	\
+				   FI_SEND | FI_RECV)
+
+#define OFI_VERBS_UD_CAPS OFI_VERBS_UD_PRIMARY_CAPS
+
+#define VERBS_UD_MSG_IOV_LIMIT	256
+#define VERBS_UD_DEF_CQ_SIZE	1024
+
+#define OFI_VERBS_UD_MR_IOV_LIMIT	1
+
+#define VERBS_IB_FABRIC_PREFIX	"IB-0x"
+#define VERBS_IWARP_FABRIC	"Ethernet-iWARP"
+#define VERBS_ANY_FABRIC	"RDMA-Fabric"
+
+#define OFI_VERBS_UD_ANY_SERVICE	0
+
+/*
+ * useful macros
+ */
+#ifndef likely
+#define likely(x) __builtin_expect((x), 1)
+#endif
+#ifndef unlikely
+#define unlikely(x) __builtin_expect((x), 0)
+#endif
+
+/* Helper macro */
+#define OFI_FID_OPS(_fid) (*_fid)->fid.ops
+#define OFI_OPS(_fid) (*_fid)->ops
+
+#define OFI_VERBS_UD_GET_INSTANCE_OF_OBJ(ofi_obj_name,				\
+					 inst)					\
+do {										\
+	inst = container_of(ofi_obj_name##_fid,					\
+			    struct ofi_verbs_ud_##ofi_obj_name,			\
+ 			    util_##ofi_obj_name.ofi_obj_name##_fid);		\
+} while (0)
+
+#define OFI_VERBS_UD_GET_INSTANCE_OF(ofi_obj_name,				\
+				     inst)					\
+do {										\
+	inst = container_of(ofi_obj_name##_fid,					\
+			    struct ofi_verbs_ud_##ofi_obj_name,			\
+ 			    util_##ofi_obj_name.ofi_obj_name##_fid.fid);	\
+} while (0)
```

Comment:
The above macros are kind of odd in code.  It would be better to have them return container_of, rather than passing in the inst parameter and setting it.  I'm also not sure about the way it wants to concatenate "_fid" to the end of the ofi_obj_name.  That forces variables to follow a hidden naming convention.  I'm not sure what the best fix is for this.